src/bots/pnlSettler.ts

```diff
@@ -0,0 +1,179 @@
+		);
+		await this.userMap.fetchAllUsers();
+	}
+
+	public reset(): void {
+		for (const intervalId of this.intervalIds) {
+			clearInterval(intervalId);
+		}
+		this.intervalIds = [];
+		delete this.userMap;
+	}
+
+	public async startIntervalLoop(intervalMs: number): Promise<void> {
+		logger.info(`${this.name} Bot started!`);
+		this.trySettlePnl();
+		const intervalId = setInterval(this.trySettlePnl.bind(this), intervalMs);
+		this.intervalIds.push(intervalId);
+	}
+
+	public async trigger(_record: any): Promise<void> {
+		return undefined;
+	}
+
+	public viewDlob(): undefined {
+		return undefined;
+	}
+
+	private async trySettlePnl() {
+		try {
+			const marketAndOracleData: {
+				[marketIndex: number]: {
+					marketAccount: MarketAccount;
+					oraclePriceData: OraclePriceData;
+				};
+			} = {};
+
+			this.markets.forEach((market) => {
+				marketAndOracleData[market.marketIndex.toNumber()] = {
+					marketAccount: this.clearingHouse.getMarketAccount(
+						market.marketIndex
+					),
+					oraclePriceData: this.clearingHouse.getOracleDataForMarket(
+						market.marketIndex
+					),
+				};
+			});
+
+			const usersToSettle: SettlePnlIxParams[] = [];
+
+			for (const user of this.userMap.values()) {
+				const userAccount = user.getUserAccount();
+
+				this.clearingHouse.fetchAccounts();
+				this.clearingHouse.getUser().fetchAccounts();
+
+				for (const settleePosition of userAccount.positions) {
+					const marketIndexNum = settleePosition.marketIndex.toNumber();
+					const unsettledPnl = calculateUnsettledPnl(
+						marketAndOracleData[marketIndexNum].marketAccount,
+						settleePosition,
+						marketAndOracleData[marketIndexNum].oraclePriceData
+					);
+					// only settle for $10 or more negative pnl
+					if (unsettledPnl.lte(MIN_PNL_TO_SETTLE)) {
+						const userData = {
+							settleeUserAccountPublicKey: user.getUserAccountPublicKey(),
+							settleeUserAccount: userAccount,
+						};
+						if (
+							usersToSettle
+								.map((item) => item.marketIndex.toNumber())
+								.includes(marketIndexNum)
+						) {
+							usersToSettle
+								.find((item) => item.marketIndex.toNumber() == marketIndexNum)
+								.users.push(userData);
+						} else {
+							usersToSettle.push({
+								users: [userData],
+								marketIndex: settleePosition.marketIndex,
+							});
+						}
+					}
+				}
+			}
+
+			usersToSettle.forEach((params) => {
+				const marketStr = this.markets.find((mkt) =>
+					mkt.marketIndex.eq(params.marketIndex)
+				).symbol;
+
+				logger.info(
+					`Trying to settle PNL for ${params.users.length} users on market ${marketStr}`
+				);
+				this.clearingHouse
+					.settlePNLs(params.users, params.marketIndex)
+					.then((txSig) => {
+						logger.info(
+							`PNL settled successfully on ${marketStr}. TxSig: ${txSig}`
+						);
+						params.users.forEach((settledUser) => {
+							this.metrics?.recordSettlePnl(
+								settledUser.settleeUserAccountPublicKey,
```

Comment:
i think you should remove the pubkey field from this metric, since [prometheus recommends avoiding high cardinality labels](https://prometheus.io/docs/practices/naming/#labels), probably just counting the number of settles done is good enough, if that number goes up a lot then we'll know something's up with user positions.

then you can also get rid of the `forEach` and pass in number of settled users to the counter instead